// proton-native/bin/utils/createElement.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var components_1 = require("../components/");
// Creates an element with an element type, props and a root instance
function createElement(type, props) {
    var COMPONENTS = {
        ROOT: function () { return components_1.Root(props); },
        APP: function () { return components_1.App(props); },
        VIEW: function () { return components_1.View(props); },
        WINDOW: function () { return components_1.Window(props); },
        VIRTUALTEXT: function () { return components_1.VirtualText(props); },
        ROOTTEXT: function () { return components_1.RootText(props); },
        IMAGE: function () { return components_1.Image(props); },
        TEXTINPUT: function () { return components_1.TextInput(props); },
        PICKERINTERNAL: function () { return components_1.PickerInternal(props); },
        BUTTON: function () { return components_1.Button(props); },
        default: undefined
    };
    return COMPONENTS[type]() || COMPONENTS.default;
}
exports.createElement = createElement;
