// proton-native/bin/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var render_1 = require("./render");
exports.AppRegistry = render_1.default;
var react_components_1 = require("./react-components");
exports.Text = react_components_1.Text;
exports.TouchableWithoutFeedback = react_components_1.TouchableWithoutFeedback;
exports.TouchableOpacity = react_components_1.TouchableOpacity;
exports.TouchableHighlight = react_components_1.TouchableHighlight;
exports.Picker = react_components_1.Picker;
var misc_1 = require("./misc");
exports.StyleSheet = misc_1.StyleSheet;
var backends_1 = require("./backends");
exports.setBackend = backends_1.setBackend;
require("./utils/requireImpl");
// Aliases for createElement method
var Window = "WINDOW";
exports.Window = Window;
var View = "VIEW";
exports.View = View;
var App = "APP";
exports.App = App;
var VirtualText = "VIRTUALTEXT";
exports.VirtualText = VirtualText;
var RootText = "ROOTTEXT";
exports.RootText = RootText;
var Image = "IMAGE";
exports.Image = Image;
var TextInput = "TEXTINPUT";
exports.TextInput = TextInput;
var PickerInternal = "PICKERINTERNAL";
exports.PickerInternal = PickerInternal;
var Button = "BUTTON";
exports.Button = Button;
