// proton-native/bin/react-components/TouchableOpacity.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var __1 = require("..");
var PropTypes = require("prop-types");
var Color = require("color");
var TouchableOpacity = /** @class */ (function (_super) {
    __extends(TouchableOpacity, _super);
    function TouchableOpacity() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        // TODO: Add animation
        _this.pressedDown = false;
        _this.startedTime = 0;
        _this.state = {
            backgroundColor: _this._getChildStyleOpacityWithDefault()
        };
        return _this;
    }
    TouchableOpacity.prototype._getChildStyleOpacityWithDefault = function () {
        return this.props.style.backgroundColor;
    };
    TouchableOpacity.prototype.onDown = function () {
        this.pressedDown = true;
        this.startedTime = new Date().getTime();
        this.setState({
            backgroundColor: Color(this.props.style.backgroundColor)
                .alpha(this.props.activeOpacity)
                .string()
        });
    };
    TouchableOpacity.prototype.onUp = function () {
        if (this.pressedDown) {
            var diff = new Date().getTime() - this.startedTime;
            if (diff > 500) {
                this.props.onLongPress
                    ? this.props.onLongPress()
                    : this.props.onPress();
            }
            else {
                this.props.onPress();
            }
            this.pressedDown = false;
            this.setState({ backgroundColor: this.props.style.backgroundColor });
        }
    };
    TouchableOpacity.prototype.render = function () {
        var _this = this;
        return (React.createElement(__1.View, __assign({}, this.props, { style: __assign(__assign({}, this.props.style), this.state), onResponderGrant: function () { return _this.onDown(); }, onResponderRelease: function () { return _this.onUp(); } }), this.props.children));
    };
    TouchableOpacity.defaultProps = {
        activeOpacity: 0.2,
        style: {},
        onPress: function () { },
        onLongPress: function () { }
    };
    TouchableOpacity.propTypes = {
        activeOpacity: PropTypes.number,
        style: PropTypes.object,
        onPress: PropTypes.func,
        onLongPress: PropTypes.func
    };
    return TouchableOpacity;
}(React.Component));
exports.default = TouchableOpacity;
